fuse_optimizers/src/fixed_lag_smoother.cpp

```diff
@@ -230,33 +230,48 @@ void FixedLagSmoother::processQueue(fuse_core::Transaction& transaction)
   auto current_time = ros::Time(0, 0);
   if (!pending_transactions_.empty())
   {
-    current_time = pending_transactions_.front_key();
+    current_time = pending_transactions_.front().stamp();
   }
   // Attempt to process each pending transaction
-  while (!pending_transactions_.empty())
+  auto sensor_blacklist = std::vector<std::string>();
+  auto transaction_iter = pending_transactions_.rbegin();
+  while (transaction_iter != pending_transactions_.rend())
   {
-    auto& element = pending_transactions_.back_value();
-    // Apply the motion models to the transaction
-    if (!applyMotionModels(element.sensor_name, *element.transaction))
+    auto& element = *transaction_iter;
+    if (std::find(sensor_blacklist.begin(), sensor_blacklist.end(), element.sensor_name) != sensor_blacklist.end())
     {
+      // We should not process transactions from this sensor
+      ++transaction_iter;
+    }
+    else if (applyMotionModels(element.sensor_name, *element.transaction))
+    {
+      // Processing was successful. Add the results to the final transaction, delete this one, and move to the next.
+      transaction.merge(*element.transaction, true);
+      ++transaction_iter;
+      pending_transactions_.erase(transaction_iter.base());  // Reverse iterators are weird
+    }
+    else
+    {
+      // The motion model processing failed.
+      // Check the transaction timeout to determine if it should be removed or skipped.
       if (element.transaction->stamp() + params_.transaction_timeout < current_time)
       {
         // Warn that this transaction has expired, then skip it.
-        ROS_ERROR_STREAM("The queued transaction with timestamp " << element.transaction->stamp()
-                          << " could not be processed after " << (current_time - element.transaction->stamp())
-                          << " seconds, which is greater than the 'transaction_timeout' value of "
-                          << params_.transaction_timeout << ". Ignoring this transaction.");
-        pending_transactions_.pop_back();
-        continue;
+        ROS_ERROR_STREAM("The queued transaction with timestamp " << element.transaction->stamp() <<
+                          " from sensor " << element.sensor_name << " could not be processed after " <<
+                          (current_time - element.transaction->stamp()) << " seconds, which is greater " <<
+                          "than the 'transaction_timeout' value of " << params_.transaction_timeout <<
+                          ". Ignoring this transaction.");
+        ++transaction_iter;
+        pending_transactions_.erase(transaction_iter.base());  // Reverse iterators are weird
       }
       else
       {
-        // The motion model failed. Stop further processing and try again next time.
-        break;
+        // The motion model failed. Stop further processing of this sensor and try again next time.
```

Comment:
Just thinking through the cases here. We're iterating in reverse, so if we just happen to have some 'bad' transaction the the motion model didn't like, it will fail to process it at every cycle, which is fine. Eventually, it will be so old that the transaction will expire and be removed. Newer measurements from that same sensor will be processed first in future iterations, so we're not blocking new data from that sensor form being processed. If the measurement was just too new, it'll get picked up in a future iteration.

Reply:
So I get that we're blacklisting single sensors, but I guess what concerns me is a 'bad' measurement from sensor A prevents all future measurements from it until it gets dropped from the queue. But it's much better than the current setup.

Reply:
I'm taking the "it's much better than the current setup" as a win for now. We can discuss a more advanced system for handling failures in the future.